cpp/src/bitmask/valid_if.cuh

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */ 
+
+#ifndef __BITMASK_VALID_IF_CUH__
+#define __BITMASK_VALID_IF_CUH__
+
+#include <cudf/types.h>
+#include <utilities/error_utils.hpp>
+#include <utilities/bit_util.cuh>
+#include <utilities/cuda_utils.hpp>
+#include <bitmask/legacy/bit_mask.cuh>
+
+#include <cub/cub.cuh>
+
+using bit_mask::bit_mask_t;
+
+namespace cudf {
+
+namespace detail {
+
+constexpr int warp_size = 32;
+
+/**
+ * @brief for each warp in the block do a reduction (summation) of the
+ * `__popc(bit_mask)` on a certain lane (default is lane 0).
+ * @param[in] bit_mask The bit_mask to be reduced.
+ * @return[out] result of each block is returned in thread 0.
+ */
+template <class bit_container, int lane = 0>
+__device__ __inline__ gdf_size_type single_lane_popc_block_reduce(bit_container bit_mask) {
+  
+  static __shared__ gdf_size_type smem[warp_size];
```

Comment:
`smem` is a generic name, but this memory is used for only one thing: warp counts, so make the name reflect the contents.

```suggestion
  static __shared__ gdf_size_type warp_counts[warp_size];
```